File: day_13/main.py
def draw_points(points, max_x, max_y):
    origami = []

    for y in range(max_y + 1):
        origami.append('\n')

        for x in range(max_x + 1):
            origami.append('#' if (x, y) in points else '.')

    return ''.join(origami)


def parse_data(f, points, folds, length):
    while (line := f.readline().strip()):
        points.add(map(int, line.split(',')))

    for line in f.readlines():
        axis, fold = line[length:length + 1], int(line[length + 2:])
        folds.append((1 if axis == 'x' else - 1) * fold)


def main():
    points, folds = set(), []

    with open('input.txt') as f:
        parse_data(f, points, folds, len('fold along '))

    for i, fold in enumerate(folds):
        new_points = set()

        for x, y in points:
            x = (2 * fold) - x if fold > 0 and x > fold else x
            y = (-2 * fold) - y if fold < 0 and y > -fold else y
            new_points.add((x, y))

        points = new_points

        if i == 0:
            print(f'Number of points after first fold (1): {len(points)}')

    x, y = map(lambda i: map(lambda p, i=i: p[i], points), [0, 1])

    print('Hidden message (2):', draw_points(points, max(x), max(y)))


if __name__ == '__main__':
    main()

File: day_13/main.rb
#!/usr/bin/env ruby

require 'set'

points = Set[]
folds = []
length = 'fold along '.length

File.open('input.txt') do |f|
  x = f.readline.strip

  until x.empty?
    points.add(x)
    x = f.readline.strip
  end

  f.read.each_line do |line|
    axis = line[length, 1]
    fold = line[length + 2..].to_i
    folds.push((axis == 'x' ? 1 : -1) * fold)
  end
end

folds.each_with_index do |fold, i|
  new_points = Set[]

  points.each do |p|
    x, y = p.split(',').map(&:to_i)
    x = fold.positive? && x > fold ? (2 * fold) - x : x
    y = fold.negative? && y > -fold ? (-2 * fold) - y : y
    new_points.add("#{x},#{y}")
  end

  points = new_points
  puts "Number of points after first fold (1): #{points.length}" if i.zero?
end

x = points.map { |p| p.split(',').first.to_i }
y = points.map { |p| p.split(',').last.to_i }

def draw_points(points, max_x, max_y)
  origami = ''

  (0..max_y).each do |y|
    (0..max_x).each do |x|
      origami += points.include?("#{x},#{y}") ? '#' : '.'
    end

    origami += "\n"
  end

  origami
end

puts 'Hidden message (2):', draw_points(points, x.max, y.max)
